vision/convmixer_cifar10/convmixer.jl

```diff
@@ -51,89 +58,75 @@ function get_data(batchsize; dataset = MLDatasets.CIFAR10, idxs = nothing)
     return train_loader, test_loader
 end
 
-function create_loss_function(dataloader, device)
-
-    function loss(model)
-        n = 0
-        l = 0.0f0
-        acc = 0.0f0
-
-        for (x,y) in dataloader
-            x,y = x |> device, y |> device
-            z = model(x)        
-            l += Flux.logitcrossentropy(z, y, agg=sum)
-            acc += sum(onecold(z).==onecold(y))
-            n += size(x)[end]
-        end
-        l / n, acc / n
+"""
+This function is used only for logging, with either train_loader or test_loader.
+"""
+function loss_and_accuracy(model, dataloader)
+    n = 0      # count images seen
+    l = 0.0f0  # total loss
+    c = 0      # count correct classifications
+
+    for (x1, y1) in dataloader
+        x, y = gpu(x1), gpu(y1)
+        z = model(x)
+        l += logitcrossentropy(z, y, agg=sum)
+        c += sum(onecold(z).==onecold(y))
+        n += size(x)[end]
     end
-
-    return loss
-   
+    return l/n, c/n  # mean loss, and accuracy
 end
 
+"""
+Main function: loads data, creates model, trains it, then saves trained model in "model.bson"
+Keyword `images=99` will use only the first 99 images of the training set (to check that it 
+runs).
+"""
+function train(; epochs=100, images=:)
+    train_loader, test_loader = get_data(128; images)
 
-function train(n_epochs=100)
-
-    #params: warning, the training can be long with these params
-    train_loader, test_loader = get_data(128)
-    η = 3e-4
+    # hyper-parameters. Note that training can be long with these!
     in_channel = 3
     patch_size = 2
     kernel_size = 7
     dim = 128
-    dimPL = 2
     depth = 18
-    use_cuda = true
-
-    #logging the losses
-    train_save = zeros(n_epochs, 2)
-    test_save = zeros(n_epochs, 2)
+    η = 3f-4  # learning rate
 
-    if use_cuda
-        device = gpu
-        @info "Training on GPU"
-    else
-        device = cpu
-        @info "Training on CPU"
-    end
-
-    train_loss_fn = create_loss_function(train_loader, device)
-    test_loss_fn = create_loss_function(test_loader, device)
-
-    model = ConvMixer(in_channel, kernel_size, patch_size, dim, depth, 10) |> device
+    model = ConvMixer(in_channel, kernel_size, patch_size, dim, depth, 10) |> gpu
 
-    ps = params(model)
-    opt = Optimiser(
+    opt = OptimiserChain(
             WeightDecay(1f-3), 
-            ClipNorm(1.0),
-            ADAM(η)
+            ClipNorm(1f0),
+            Adam(η),
             )
```

Comment:
This exposes a real error: composing `ClipNorm` after another optimiser means it tries to take `norm(::Broadcasted)` which fails:
```
julia> train(epochs=10, images=128)
ERROR: Scalar indexing is disallowed.
...
Stacktrace:
...
 [12] norm(itr::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{4}, NTuple{4, Base.OneTo{Int64}}, typeof(+), Tuple{CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}, Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{4}, Nothing, typeof(*), Tuple{Float32, CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}}}}}, p::Float32)
    @ LinearAlgebra ~/julia-9ded051e9f/share/julia/stdlib/v1.10/LinearAlgebra/src/generic.jl:596
 [13] apply!(o::Optimisers.ClipNorm{Float32}, state::Nothing, x::CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}, dx::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{4}, NTuple{4, Base.OneTo{Int64}}, typeof(+), Tuple{CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}, Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{4}, Nothing, typeof(*), Tuple{Float32, CuArray{Float32, 4, CUDA.Mem.DeviceBuffer}}}}})
    @ Optimisers ~/.julia/packages/Optimisers/kPdJV/src/rules.jl:584
```

Reply:
https://github.com/FluxML/Optimisers.jl/issues/127